File: ios/TriCoach/Animation/Koerper.swift
import Foundation
import simd

/// Eine Pose: Gelenkwinkel in Grad und Beckenlage in Metern, wie sie das
/// Backend schreibt („huefte_beugen_l“: 52). Fehlende Werte sind 0 — außer
/// der Beckenhöhe (`Koerper.vorgaben`).
typealias Pose = [String: Double]
typealias Punkt = SIMD3<Double>

/// Das Körpermodell der Übungsanimationen — Zahl für Zahl dasselbe wie
/// `backend/app/animation/koerper.py`.
///
/// Der Löser im Backend hat jede Pose für **dieses** Modell gerechnet: dass die
/// Ferse aufliegt und die Hand am Knöchel ist, stimmt nur, solange Maße,
/// Achsen und Reihenfolge der Drehungen hier dieselben sind. Weicht die App
/// ab, schweben Hände über der Matte. Deshalb prüft der Simulator-Debug-Build
/// beim Start gegen `backend/tests/fixtures/animation_fk_referenz.json` —
/// dieselbe Datei, gegen die `test_animation.py` die Python-Seite prüft.
///
/// Koordinaten: x nach links der Figur, y nach oben, z nach vorn, Boden bei
/// y = 0.
enum Koerper {
    enum Mass {
        static let beckenBreite = 0.10
        static let wirbel = 0.46
        static let hals = 0.10
        static let kopf = 0.115
        static let schulterBreite = 0.19
        static let oberarm = 0.29
        static let unterarm = 0.26
        static let hand = 0.08
        static let oberschenkel = 0.44
        static let unterschenkel = 0.43
        static let fuss = 0.19
    }

    /// Ohne Angabe steht das Becken auf Standhöhe — die Sohlen genau auf dem Boden.
    static let vorgaben: Pose = ["y": 0.973]

    struct Segment {
        let von: String
        let nach: String
        let radius: Double
        /// Die Muskelgruppe, die eine Animation unter `betont` hervorheben kann.
        let gruppe: String

        /// `l`/`r`, wenn beide Enden auf derselben Seite liegen.
        var seite: String? {
            guard let a = von.last, von.dropLast().hasSuffix("_"), nach.hasSuffix("_\(a)"), a == "l" || a == "r" else {
                return nil
            }
            return String(a)
        }
    }

    static let segmente: [Segment] = {
        var liste = [
            Segment(von: "becken", nach: "taille", radius: 0.112, gruppe: "rumpf"),
            Segment(von: "taille", nach: "brust", radius: 0.128, gruppe: "rumpf"),
            Segment(von: "schulter_l", nach: "schulter_r", radius: 0.062, gruppe: "schultern"),
            Segment(von: "huefte_l", nach: "huefte_r", radius: 0.088, gruppe: "huefte"),
            Segment(von: "gesaess_l", nach: "gesaess_r", radius: 0.07, gruppe: "gesaess"),
            Segment(von: "brust", nach: "nacken", radius: 0.05, gruppe: "hals"),
        ]
        for s in ["l", "r"] {
            liste += [
                Segment(von: "schulter_\(s)", nach: "ellbogen_\(s)", radius: 0.047, gruppe: "arme"),
                Segment(von: "ellbogen_\(s)", nach: "hand_\(s)", radius: 0.039, gruppe: "arme"),
                Segment(von: "hand_\(s)", nach: "finger_\(s)", radius: 0.033, gruppe: "arme"),
                Segment(von: "huefte_\(s)", nach: "knie_\(s)", radius: 0.078, gruppe: "oberschenkel"),
                Segment(von: "knie_\(s)", nach: "knoechel_\(s)", radius: 0.056, gruppe: "waden"),
                Segment(von: "ferse_\(s)", nach: "zehen_\(s)", radius: 0.038, gruppe: "fuesse"),
            ]
        }
        return liste
    }()

    static let kopfRadius = 0.115
    static let naseRadius = 0.028

    static func wert(_ pose: Pose, _ schluessel: String) -> Double {
        pose[schluessel] ?? vorgaben[schluessel] ?? 0
    }

    // MARK: - Vorwärtskinematik

    /// Alle Gelenkpunkte einer Pose.
    static func punkte(_ pose: Pose) -> [String: Punkt] {
        let g = { (k: String) in Koerper.wert(pose, k) }
        let r0 = ry(g("gieren")) * rx(g("nicken")) * rz(g("rollen")) * ry(g("drehen"))
        let becken = Punkt(g("x"), g("y"), g("z"))
        var p: [String: Punkt] = ["becken": becken]
        p.reserveCapacity(27)

        // Die Wirbelsäule in zwei Abschnitten, jeder mit der halben Bewegung —
        // aus einem Katzenbuckel wird so ein Bogen.
        func abschnitt(_ r: simd_double3x3) -> simd_double3x3 {
            r * rz(-g("rumpf_seit") / 2) * rx(g("rumpf_beugen") / 2) * ry(g("rumpf_drehen") / 2)
        }
        let rl = abschnitt(r0)
        let taille = becken + rl * Punkt(0, Mass.wirbel * 0.45, 0)
        let rw = abschnitt(rl)
        let brust = taille + rw * Punkt(0, Mass.wirbel * 0.55, 0)
        let rk = rw * rx(g("kopf_beugen"))
        let nacken = brust + rk * Punkt(0, Mass.hals, 0)
        let kopf = nacken + rk * Punkt(0, Mass.kopf * 0.9, 0.01)
        p["taille"] = taille
        p["brust"] = brust
        p["nacken"] = nacken
        p["kopf"] = kopf
        p["nase"] = kopf + rk * Punkt(0, -0.01, Mass.kopf * 0.95)

        for (seite, s) in [("l", 1.0), ("r", -1.0)] {
            let schulter = brust + rw * Punkt(s * Mass.schulterBreite, -0.03, 0)
            let rs = rw * gelenk(
                -g("schulter_beugen_\(seite)"), g("schulter_abspreizen_\(seite)"), g("schulter_drehen_\(seite)"), s
            )
            let ellbogen = schulter + rs * Punkt(0, -Mass.oberarm, 0)
            let re = rs * rx(-g("ellbogen_\(seite)"))
            let hand = ellbogen + re * Punkt(0, -Mass.unterarm, 0)
            let finger = hand + (re * rx(g("handgelenk_\(seite)"))) * Punkt(0, -Mass.hand, 0)

            let huefte = becken + r0 * Punkt(s * Mass.beckenBreite, -0.02, 0)
            let gesaess = becken + r0 * Punkt(s * 0.07, -0.06, -0.042)
            let rh = r0 * gelenk(
                -g("huefte_beugen_\(seite)"), g("huefte_abspreizen_\(seite)"), g("huefte_drehen_\(seite)"), s
            )
            let knie = huefte + rh * Punkt(0, -Mass.oberschenkel, 0)
            let rkn = rh * rx(g("knie_\(seite)"))
            let knoechel = knie + rkn * Punkt(0, -Mass.unterschenkel, 0)
            let rf = rkn * rx(-g("fuss_\(seite)"))

            p["schulter_\(seite)"] = schulter
            p["ellbogen_\(seite)"] = ellbogen
            p["hand_\(seite)"] = hand
            p["finger_\(seite)"] = finger
            p["huefte_\(seite)"] = huefte
            p["gesaess_\(seite)"] = gesaess
            p["knie_\(seite)"] = knie
            p["knoechel_\(seite)"] = knoechel
            p["zehen_\(seite)"] = knoechel + rf * Punkt(0, -0.045, Mass.fuss)
            p["ferse_\(seite)"] = knoechel + rf * Punkt(0, -0.045, -0.04)
        }
        return p
    }

    // MARK: - Drehungen (Zeilen wie in Python, damit der Vergleich leicht fällt)

    private static func bogen(_ grad: Double) -> Double { grad * .pi / 180 }

    private static func rx(_ grad: Double) -> simd_double3x3 {
        let c = cos(bogen(grad)), s = sin(bogen(grad))
        return simd_double3x3(rows: [Punkt(1, 0, 0), Punkt(0, c, -s), Punkt(0, s, c)])
    }

    private static func ry(_ grad: Double) -> simd_double3x3 {
        let c = cos(bogen(grad)), s = sin(bogen(grad))
        return simd_double3x3(rows: [Punkt(c, 0, s), Punkt(0, 1, 0), Punkt(-s, 0, c)])
    }

    private static func rz(_ grad: Double) -> simd_double3x3 {
        let c = cos(bogen(grad)), s = sin(bogen(grad))
        return simd_double3x3(rows: [Punkt(c, -s, 0), Punkt(s, c, 0), Punkt(0, 0, 1)])
    }

    /// Drehung um eine beliebige Achse (Rodrigues).
    private static func achseWinkel(_ achse: Punkt, _ grad: Double) -> simd_double3x3 {
        let laenge = simd_length(achse)
        guard laenge >= 1e-12, abs(grad) >= 1e-12 else { return matrix_identity_double3x3 }
        let a = achse / laenge
        let (x, y, z) = (a.x, a.y, a.z)
        let c = cos(bogen(grad)), s = sin(bogen(grad)), k = 1 - c
        return simd_double3x3(rows: [
            Punkt(c + x * x * k, x * y * k - z * s, x * z * k + y * s),
            Punkt(y * x * k + z * s, c + y * y * k, y * z * k - x * s),
            Punkt(z * x * k - y * s, z * y * k + x * s, c + z * z * k),
        ])
    }

    /// Kugelgelenk: Schwenk aus Beugen und Abspreizen **gemeinsam**, dann die
    /// Drehung um die Längsachse. Warum kein Euler-Winkel: siehe `_gelenk` in
    /// `koerper.py` und „Kugelgelenke sind Schwenkvektoren“ in
    /// `docs/animationen.md`.
    private static func gelenk(_ beugen: Double, _ abspreizen: Double, _ drehen: Double, _ seite: Double) -> simd_double3x3 {
        achseWinkel(Punkt(beugen, 0, seite * abspreizen), (beugen * beugen + abspreizen * abspreizen).squareRoot())
            * ry(seite * drehen)
    }
}

// MARK: - Abspielen

extension Koerper {
    /// Linear zwischen zwei Posen; was nur eine nennt, kommt aus den Vorgaben.
    static func zwischen(_ a: Pose, _ b: Pose, _ t: Double) -> Pose {
        var ergebnis = Pose(minimumCapacity: a.count + b.count)
        for k in Set(a.keys).union(b.keys) {
            let va = wert(a, k)
            ergebnis[k] = va + (wert(b, k) - va) * t
        }
        return ergebnis
    }

    /// Weich an- und auslaufen — über den **ganzen** Übergang, nicht je Teilstück.
    static func glatt(_ t: Double) -> Double { 0.5 - 0.5 * cos(.pi * t) }
}

extension Bewegung {
    /// Wie lange eine Schleife dauert.
    var dauer: Double { max(0.1, ablauf.reduce(0) { $0 + $1.haltenS + $1.uebergangS }) }

    /// Die Pose zum Zeitpunkt `t` (Sekunden, beliebig groß — die Schleife wiederholt sich).
    ///
    /// Übergänge laufen über die gelösten Zwischenbilder des Backends, sonst
    /// rutschte ein stehender Fuß beim Überblenden über den Boden.
    func pose(bei zeit: Double) -> Pose {
        guard let erstes = ablauf.first else { return [:] }
        var t = zeit.truncatingRemainder(dividingBy: dauer)
        if t < 0 { t += dauer }
        for (i, bild) in ablauf.enumerated() {
            if t < bild.haltenS { return bild.pose }
            t -= bild.haltenS
            if t < bild.uebergangS {
                let naechste = ablauf[(i + 1) % ablauf.count].pose
                return Self.uebergang(bild, naechste, t / bild.uebergangS)
            }
            t -= bild.uebergangS
        }
        return erstes.pose
    }

    static func uebergang(_ bild: Schluesselbild, _ naechste: Pose, _ t: Double) -> Pose {
        let kette = [bild.pose] + bild.zwischen + [naechste]
        let e = Koerper.glatt(t) * Double(kette.count - 1)
        let i = min(Int(e), kette.count - 2)
        return Koerper.zwischen(kette[i], kette[i + 1], e - Double(i))
    }
}

// MARK: - Prüfung gegen die Referenz

#if DEBUG && targetEnvironment(simulator)
extension Koerper {
    private struct Referenz: Decodable {
        struct Eintrag: Decodable {
            let pose: Pose
            let punkte: [String: [Double]]
        }
        let posen: [Eintrag]
    }

    /// Rechnet die Referenzposen des Backends nach und schreibt die größte
    /// Abweichung ins Diagnoseprotokoll. Über einem Zehntelmillimeter ist das
    /// Modell auseinandergelaufen.
    static func pruefeReferenz() {
        let datei = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()  // Animation
            .deletingLastPathComponent()  // TriCoach
            .deletingLastPathComponent()  // ios
            .deletingLastPathComponent()  // Repo
            .appendingPathComponent("backend/tests/fixtures/animation_fk_referenz.json")
        guard let daten = try? Data(contentsOf: datei),
              let referenz = try? JSONDecoder().decode(Referenz.self, from: daten)
        else {
            protokolliere("Körpermodell: Referenz nicht lesbar (\(datei.path))")
            return
        }
        var groesste = 0.0
        var wo = ""
        for eintrag in referenz.posen {
            let ist = punkte(eintrag.pose)
            for (gelenk, soll) in eintrag.punkte {
                guard let p = ist[gelenk], soll.count == 3 else {
                    groesste = .infinity
                    wo = "\(gelenk) fehlt"
                    continue
                }
                let abweichung = simd_length(p - Punkt(soll[0], soll[1], soll[2]))
                if abweichung > groesste {
                    groesste = abweichung
                    wo = gelenk
                }
            }
        }
        let urteil = groesste < 1e-4 ? "stimmt" : "WEICHT AB"
        protokolliere(
            "Körpermodell \(urteil): \(referenz.posen.count) Referenzposen, größte Abweichung "
                + String(format: "%.2e", groesste) + " m (\(wo))"
        )
    }
}
#endif

File: backend/app/animation/koerper.py
"""Das Körpermodell: Gelenkwinkel rein, Gelenkpunkte im Raum raus.

Ein einziges Modell für alles, was eine Animation berührt: den Löser, der
Kontaktpunkte auf den Boden bringt, die Prüfung einer KI-Antwort und — Zeile
für Zeile nachgebaut — den Renderer der iOS-App (`Figur.swift`). Weichen die
beiden voneinander ab, zeigt die App eine andere Haltung als die, die hier
gelöst wurde. `tests/test_animation.py` hält deshalb Referenzpunkte fest, gegen
die auch die Swift-Seite geprüft wird.

**Konventionen** (Ruhehaltung: stehend, Arme hängend, Blick nach +z):

* Welt: x nach links der Figur, y nach oben, z nach vorn. Meter, Boden y = 0.
* Wurzel ist das Becken: `x`, `y`, `z` und die Lage `gieren` (um die
  Senkrechte der Welt), `nicken` (um x), `rollen` (um z) und `drehen` (um die
  eigene Längsachse) — angewandt als Ry(gieren)·Rx(nicken)·Rz(rollen)·Ry(drehen).
  `nicken = -90` ist die Rückenlage (Kopf nach -z, Gesicht nach oben),
  `nicken = +90` die Bauch- bzw. Vierfüßlerlage (Kopf nach +z, Gesicht nach
  unten), `rollen = 90` die Seitenlage auf der rechten Seite. `drehen` rollt
  einen liegenden Körper zur Seite, ohne ihn auf dem Boden zu verdrehen.
* Gelenkwinkel in Grad, positiv in der anatomisch üblichen Richtung:
  Beugen nach vorn, Abspreizen vom Körper weg, Außendrehung nach außen — für
  beide Seiten gleich, die Spiegelung erledigt das Modell.

Reines Python ohne numpy: Das Modul läuft im Add-on auf dem Raspberry Pi, und
eine Abhängigkeit mit C-Erweiterung verlängerte dort den Build um Minuten.
"""

from __future__ import annotations

import math
from typing import Mapping

Vek = tuple[float, float, float]
Mat = tuple[Vek, Vek, Vek]

# Körpermaße in Metern — ein Erwachsener von gut 1,75 m.
MASSE = {
    "becken_breite": 0.10,
    "wirbel": 0.46,
    "hals": 0.10,
    "kopf": 0.115,
    "schulter_breite": 0.19,
    "oberarm": 0.29,
    "unterarm": 0.26,
    "hand": 0.08,
    "oberschenkel": 0.44,
    "unterschenkel": 0.43,
    "fuss": 0.19,
}

SEITEN = ("l", "r")

# Jeder Posenschlüssel mit zulässigem Bereich. Was hier nicht steht, gibt es
# nicht — eine KI-Antwort mit erfundenem Gelenk wird abgelehnt, statt still
# ignoriert zu werden. Die Bereiche sind großzügig (Yoga), aber nicht beliebig:
# Ein Knie mit 250° ist ein Tippfehler, keine Übung.
_WURZEL = {
    "x": (-3.0, 3.0), "y": (-0.5, 2.5), "z": (-3.0, 3.0),
    "gieren": (-360.0, 360.0), "nicken": (-360.0, 360.0), "rollen": (-360.0, 360.0),
    "drehen": (-360.0, 360.0),
}
_RUMPF = {
    "rumpf_beugen": (-60.0, 120.0), "rumpf_seit": (-60.0, 60.0),
    "rumpf_drehen": (-90.0, 90.0), "kopf_beugen": (-60.0, 80.0),
    "atmen": (0.0, 1.0),
}
_JE_SEITE = {
    "schulter_beugen": (-90.0, 200.0), "schulter_abspreizen": (-90.0, 200.0),
    "schulter_drehen": (-100.0, 100.0), "ellbogen": (0.0, 160.0),
    "handgelenk": (-80.0, 100.0),
    "huefte_beugen": (-60.0, 160.0), "huefte_abspreizen": (-50.0, 110.0),
    "huefte_drehen": (-80.0, 80.0), "knie": (0.0, 165.0), "fuss": (-60.0, 50.0),
}
BEREICHE: dict[str, tuple[float, float]] = {
    **_WURZEL,
    **_RUMPF,
    **{f"{k}_{s}": v for k, v in _JE_SEITE.items() for s in SEITEN},
}

# Ohne Angabe: Becken auf Standhöhe (Fußsohle genau auf dem Boden), alles andere 0.
VORGABEN = {"y": 0.973}

# Segmente für Renderer und Schatten: (von, nach, Radius, Muskelgruppe).
# Die Gruppe ist das, was eine Animation unter `betont` hervorheben kann — mit
# Seitenzusatz (`oberschenkel_l`) auch nur eine Seite.
SEGMENTE: tuple[tuple[str, str, float, str], ...] = (
    ("becken", "taille", 0.112, "rumpf"),
    ("taille", "brust", 0.128, "rumpf"),
    ("schulter_l", "schulter_r", 0.062, "schultern"),
    ("huefte_l", "huefte_r", 0.088, "huefte"),
    ("gesaess_l", "gesaess_r", 0.07, "gesaess"),
    ("brust", "nacken", 0.05, "hals"),
    *(
        (f"{a}_{s}", f"{b}_{s}", r, g)
        for s in SEITEN
        for a, b, r, g in (
            ("schulter", "ellbogen", 0.047, "arme"),
            ("ellbogen", "hand", 0.039, "arme"),
            ("hand", "finger", 0.033, "arme"),
            ("huefte", "knie", 0.078, "oberschenkel"),
            ("knie", "knoechel", 0.056, "waden"),
            ("ferse", "zehen", 0.038, "fuesse"),
        )
    ),
)
KOPF_RADIUS = 0.115
NASE_RADIUS = 0.028

MUSKELGRUPPEN = frozenset(g for *_, g in SEGMENTE) | {"kopf"}

# Wie hoch ein Gelenk über dem Boden steht, wenn es aufliegt: der Radius des
# Segments, das dort den Boden berührt.
KONTAKT: dict[str, float] = {
    "kopf": KOPF_RADIUS, "nacken": 0.05, "brust": 0.128, "taille": 0.12,
    "becken": 0.112,
    **{
        f"{g}_{s}": r
        for s in SEITEN
        for g, r in (
            ("schulter", 0.062), ("ellbogen", 0.047), ("hand", 0.039),
            ("finger", 0.033), ("huefte", 0.088), ("gesaess", 0.07),
            ("knie", 0.06), ("knoechel", 0.056), ("ferse", 0.038),
            ("zehen", 0.038),
        )
    },
}
GELENKE = frozenset(KONTAKT) | {"nase"}


# --------------------------------------------------------------------------
# Kleine Linearalgebra
# --------------------------------------------------------------------------


def _rx(grad: float) -> Mat:
    t = math.radians(grad)
    c, s = math.cos(t), math.sin(t)
    return ((1.0, 0.0, 0.0), (0.0, c, -s), (0.0, s, c))


def _ry(grad: float) -> Mat:
    t = math.radians(grad)
    c, s = math.cos(t), math.sin(t)
    return ((c, 0.0, s), (0.0, 1.0, 0.0), (-s, 0.0, c))


def _rz(grad: float) -> Mat:
    t = math.radians(grad)
    c, s = math.cos(t), math.sin(t)
    return ((c, -s, 0.0), (s, c, 0.0), (0.0, 0.0, 1.0))


def _mm(a: Mat, b: Mat) -> Mat:
    return tuple(
        tuple(sum(a[i][k] * b[k][j] for k in range(3)) for j in range(3))
        for i in range(3)
    )  # type: ignore[return-value]


def _mv(a: Mat, v: Vek) -> Vek:
    return (
        a[0][0] * v[0] + a[0][1] * v[1] + a[0][2] * v[2],
        a[1][0] * v[0] + a[1][1] * v[1] + a[1][2] * v[2],
        a[2][0] * v[0] + a[2][1] * v[1] + a[2][2] * v[2],
    )


def _plus(a: Vek, b: Vek) -> Vek:
    return (a[0] + b[0], a[1] + b[1], a[2] + b[2])


def _achse_winkel(achse: Vek, grad: float) -> Mat:
    """Drehung um eine beliebige Achse (Rodrigues)."""
    laenge = math.sqrt(achse[0] ** 2 + achse[1] ** 2 + achse[2] ** 2)
    if laenge < 1e-12 or abs(grad) < 1e-12:
        return ((1.0, 0.0, 0.0), (0.0, 1.0, 0.0), (0.0, 0.0, 1.0))
    x, y, z = (a / laenge for a in achse)
    t = math.radians(grad)
    c, s, k = math.cos(t), math.sin(t), 1 - math.cos(t)
    return (
        (c + x * x * k, x * y * k - z * s, x * z * k + y * s),
        (y * x * k + z * s, c + y * y * k, y * z * k - x * s),
        (z * x * k - y * s, z * y * k + x * s, c + z * z * k),
    )


def _gelenk(beugen: float, abspreizen: float, drehen: float, seite: int) -> Mat:
    """Kugelgelenk (Schulter, Hüfte): Schwenk aus Beugen und Abspreizen, dann Drehung.

    Beugen und Abspreizen bilden **gemeinsam** einen Drehvektor — nicht zwei
    nacheinander ausgeführte Drehungen. Hintereinander ausgeführt gäbe es eine
    tote Stellung: Bei 90° Beugen zeigt der Arm nach vorn, genau entlang der
    Achse, um die das Abspreizen danach dreht, und keine Abspreizung bewegte
    ihn mehr zur Seite („Thread the Needle“ war damit nicht darstellbar). Als
    Drehvektor ist jede Kombination erreichbar, und für eine reine Beugung oder
    reine Abspreizung ändert sich nichts.

    `beugen` kommt mit dem Vorzeichen, das die Drehung um x braucht; `seite`
    spiegelt Abspreizen und Drehen, damit ein positiver Wert links wie rechts
    vom Körper weg bzw. nach außen zeigt.
    """
    schwenk = _achse_winkel(
        (beugen, 0.0, seite * abspreizen),
        math.sqrt(beugen * beugen + abspreizen * abspreizen),
    )
    return _mm(schwenk, _ry(seite * drehen))


def wert(pose: Mapping[str, float], schluessel: str) -> float:
    return float(pose.get(schluessel, VORGABEN.get(schluessel, 0.0)))


def punkte(pose: Mapping[str, float]) -> dict[str, Vek]:
    """Alle Gelenkpunkte einer Pose (Vorwärtskinematik)."""
    g = lambda k: wert(pose, k)  # noqa: E731
    r0 = _mm(_mm(_mm(_ry(g("gieren")), _rx(g("nicken"))), _rz(g("rollen"))), _ry(g("drehen")))
    becken: Vek = (g("x"), g("y"), g("z"))
    p: dict[str, Vek] = {"becken": becken}

    # Wirbelsäule in zwei Abschnitten (Lende bis Taille, Brust bis Schultern),
    # jeder trägt die Hälfte der Bewegung — so wird aus einem Katzenbuckel ein
    # Bogen und kein abgeknickter Stab. Vorbeugen = Rx(+), zur linken Seite
    # neigen = Rz(-), drehen um die eigene Achse.
    def _abschnitt(r: Mat) -> Mat:
        return _mm(_mm(_mm(r, _rz(-g("rumpf_seit") / 2)), _rx(g("rumpf_beugen") / 2)), _ry(g("rumpf_drehen") / 2))

    rl = _abschnitt(r0)
    p["taille"] = _plus(becken, _mv(rl, (0.0, MASSE["wirbel"] * 0.45, 0.0)))
    rw = _abschnitt(rl)
    brust = _plus(p["taille"], _mv(rw, (0.0, MASSE["wirbel"] * 0.55, 0.0)))
    p["brust"] = brust
    rk = _mm(rw, _rx(g("kopf_beugen")))
    p["nacken"] = _plus(brust, _mv(rk, (0.0, MASSE["hals"], 0.0)))
    p["kopf"] = _plus(p["nacken"], _mv(rk, (0.0, MASSE["kopf"] * 0.9, 0.01)))
    p["nase"] = _plus(p["kopf"], _mv(rk, (0.0, -0.01, MASSE["kopf"] * 0.95)))

    for seite, s in (("l", 1), ("r", -1)):
        schulter = _plus(brust, _mv(rw, (s * MASSE["schulter_breite"], -0.03, 0.0)))
        rs = _mm(rw, _gelenk(
            -g(f"schulter_beugen_{seite}"), g(f"schulter_abspreizen_{seite}"),
            g(f"schulter_drehen_{seite}"), s,
        ))
        ellbogen = _plus(schulter, _mv(rs, (0.0, -MASSE["oberarm"], 0.0)))
        re = _mm(rs, _rx(-g(f"ellbogen_{seite}")))
        hand = _plus(ellbogen, _mv(re, (0.0, -MASSE["unterarm"], 0.0)))
        # Handgelenk strecken (positiv) klappt die Finger zum Handrücken hin —
        # so liegt im Stütz die Handfläche flach auf.
        rhg = _mm(re, _rx(g(f"handgelenk_{seite}")))
        finger = _plus(hand, _mv(rhg, (0.0, -MASSE["hand"], 0.0)))

        huefte = _plus(becken, _mv(r0, (s * MASSE["becken_breite"], -0.02, 0.0)))
        gesaess = _plus(becken, _mv(r0, (s * 0.07, -0.06, -0.042)))
        rh = _mm(r0, _gelenk(
            -g(f"huefte_beugen_{seite}"), g(f"huefte_abspreizen_{seite}"),
            g(f"huefte_drehen_{seite}"), s,
        ))
        knie = _plus(huefte, _mv(rh, (0.0, -MASSE["oberschenkel"], 0.0)))
        rkn = _mm(rh, _rx(g(f"knie_{seite}")))
        knoechel = _plus(knie, _mv(rkn, (0.0, -MASSE["unterschenkel"], 0.0)))
        rf = _mm(rkn, _rx(-g(f"fuss_{seite}")))
        zehen = _plus(knoechel, _mv(rf, (0.0, -0.045, MASSE["fuss"])))
        ferse = _plus(knoechel, _mv(rf, (0.0, -0.045, -0.04)))

        p.update({
            f"schulter_{seite}": schulter, f"ellbogen_{seite}": ellbogen,
            f"hand_{seite}": hand, f"finger_{seite}": finger,
            f"huefte_{seite}": huefte, f"gesaess_{seite}": gesaess,
            f"knie_{seite}": knie, f"knoechel_{seite}": knoechel,
            f"zehen_{seite}": zehen, f"ferse_{seite}": ferse,
        })
    return p


def tiefster_ueberstand(pose: Mapping[str, float]) -> float:
    """Wie weit der Körper in den Boden ragt (positiv = drin), in Metern.

    Gemessen an der Unterkante jedes Segments. Eine gelöste Pose sollte hier
    höchstens ein, zwei Zentimeter zeigen — mehr heißt, dass eine Hand oder ein
    Knie im Boden steckt.
    """
    p = punkte(pose)
    tiefe = -math.inf
    for a, b, r, _ in SEGMENTE:
        tiefe = max(tiefe, r - p[a][1], r - p[b][1])
    return max(tiefe, KOPF_RADIUS - p["kopf"][1])
